File: examples/numerical_propagation/covariance_rtn.py
# /// script
# dependencies = ["brahe", "numpy"]
# ///
"""
Covariance propagation in RTN (Radial-Tangential-Normal) frame.
Demonstrates frame-specific covariance retrieval and physical interpretation.
"""

import numpy as np
import brahe as bh

# Initialize EOP data
bh.initialize_eop()

# Create initial epoch and state
epoch = bh.Epoch.from_datetime(2024, 1, 1, 12, 0, 0.0, 0.0, bh.TimeSystem.UTC)
oe = np.array([bh.R_EARTH + 500e3, 0.01, 45.0, 0.0, 0.0, 0.0])
state = bh.state_koe_to_eci(oe, bh.AngleFormat.DEGREES)

# Enable STM for covariance propagation
prop_config = bh.NumericalPropagationConfig.default().with_stm().with_stm_history()

# Define initial covariance in ECI frame
# Position uncertainty: 10 m in each axis
# Velocity uncertainty: 0.01 m/s in each axis
P0 = np.diag([100.0, 100.0, 100.0, 0.0001, 0.0001, 0.0001])

# Create propagator with initial covariance
prop = bh.NumericalOrbitPropagator(
    epoch,
    state,
    prop_config,
    bh.ForceModelConfig.two_body(),
    None,
    initial_covariance=P0,
)

print("=== Covariance in RTN Frame ===\n")
print("Initial position std (ECI): 10.0 m in each axis")

# Propagate for one orbital period
orbital_period = bh.orbital_period(oe[0])
prop.propagate_to(epoch + orbital_period)

# Get covariance in different frames
target = epoch + orbital_period
P_gcrf = prop.covariance_gcrf(target)
P_rtn = prop.covariance_rtn(target)

# Extract position covariances (3x3 upper-left block)
pos_cov_gcrf = P_gcrf[:3, :3]
pos_cov_rtn = P_rtn[:3, :3]

print("\n--- GCRF Frame Results ---")
print("Position std (X, Y, Z):")
print(f"  X: {np.sqrt(pos_cov_gcrf[0, 0]):.1f} m")
print(f"  Y: {np.sqrt(pos_cov_gcrf[1, 1]):.1f} m")
print(f"  Z: {np.sqrt(pos_cov_gcrf[2, 2]):.1f} m")

print("\n--- RTN Frame Results ---")
print("Position std (R, T, N):")
print(f"  Radial (R):     {np.sqrt(pos_cov_rtn[0, 0]):.1f} m  <- Altitude uncertainty")
print(f"  Tangential (T): {np.sqrt(pos_cov_rtn[1, 1]):.1f} m  <- Along-track timing")
print(f"  Normal (N):     {np.sqrt(pos_cov_rtn[2, 2]):.1f} m  <- Cross-track offset")

# Physical interpretation
print("\n--- Physical Interpretation ---")
print("RTN frame aligns with the orbit:")
print("  R (Radial): Points from Earth center to satellite")
print("  T (Tangential): Points along velocity direction")
print("  N (Normal): Completes right-hand system (cross-track)")
print()
print("Key insight: Along-track (T) uncertainty grows fastest because")
print("velocity uncertainty causes timing errors that accumulate.")
print(
    f"After one orbit: T/R ratio = {np.sqrt(pos_cov_rtn[1, 1]) / np.sqrt(pos_cov_rtn[0, 0]):.1f}x"
)

# Show correlation structure
print("\n--- Position Correlation Matrix (RTN) ---")
pos_std_rtn = np.sqrt(np.diag(pos_cov_rtn))
corr_rtn = pos_cov_rtn / np.outer(pos_std_rtn, pos_std_rtn)
print("       R      T      N")
for i, name in enumerate(["R", "T", "N"]):
    print(
        f"  {name}  {corr_rtn[i, 0]:6.3f} {corr_rtn[i, 1]:6.3f} {corr_rtn[i, 2]:6.3f}"
    )

# Validate
assert P_gcrf.shape == (6, 6)
assert P_rtn.shape == (6, 6)
assert np.sqrt(pos_cov_rtn[1, 1]) > np.sqrt(pos_cov_rtn[0, 0])  # T > R

print("\nExample validated successfully!")

File: examples/numerical_propagation/covariance_rtn.rs
//! Covariance propagation in RTN (Radial-Tangential-Normal) frame.
//! Demonstrates frame-specific covariance retrieval and physical interpretation.

use brahe as bh;
use bh::traits::{DOrbitCovarianceProvider, DStatePropagator};
use nalgebra as na;
use std::f64::consts::PI;

fn main() {
    // Initialize EOP data
    bh::initialize_eop().unwrap();

    // Create initial epoch and state
    let epoch = bh::Epoch::from_datetime(2024, 1, 1, 12, 0, 0.0, 0.0, bh::TimeSystem::UTC);
    let oe = na::SVector::<f64, 6>::new(bh::R_EARTH + 500e3, 0.01, 45.0, 0.0, 0.0, 0.0);
    let state = bh::state_koe_to_eci(oe, bh::AngleFormat::Degrees);

    // Enable STM for covariance propagation
    let mut prop_config = bh::NumericalPropagationConfig::default();
    prop_config.variational.enable_stm = true;
    prop_config.variational.store_stm_history = true;

    // Define initial covariance in ECI frame
    // Position uncertainty: 10 m in each axis
    // Velocity uncertainty: 0.01 m/s in each axis
    let p0: na::DMatrix<f64> = na::DMatrix::from_diagonal(&na::DVector::from_vec(vec![
        100.0, 100.0, 100.0, 0.0001, 0.0001, 0.0001,
    ]));

    // Create propagator with initial covariance
    let mut prop = bh::DNumericalOrbitPropagator::new(
        epoch,
        na::DVector::from_column_slice(state.as_slice()),
        prop_config,
        bh::ForceModelConfig::two_body_gravity(),
        None,
        None,
        None,
        Some(p0),
    )
    .unwrap();

    println!("=== Covariance in RTN Frame ===\n");
    println!("Initial position std (ECI): 10.0 m in each axis");

    // Propagate for one orbital period
    let orbital_period = 2.0 * PI * (oe[0].powi(3) / bh::GM_EARTH).sqrt();
    prop.propagate_to(epoch + orbital_period);

    // Get covariance in different frames
    let target = epoch + orbital_period;
    let p_gcrf = prop.covariance_gcrf(target).unwrap();
    let p_rtn = prop.covariance_rtn(target).unwrap();

    println!("\n--- GCRF Frame Results ---");
    println!("Position std (X, Y, Z):");
    println!("  X: {:.1} m", p_gcrf[(0, 0)].sqrt());
    println!("  Y: {:.1} m", p_gcrf[(1, 1)].sqrt());
    println!("  Z: {:.1} m", p_gcrf[(2, 2)].sqrt());

    println!("\n--- RTN Frame Results ---");
    println!("Position std (R, T, N):");
    println!(
        "  Radial (R):     {:.1} m  <- Altitude uncertainty",
        p_rtn[(0, 0)].sqrt()
    );
    println!(
        "  Tangential (T): {:.1} m  <- Along-track timing",
        p_rtn[(1, 1)].sqrt()
    );
    println!(
        "  Normal (N):     {:.1} m  <- Cross-track offset",
        p_rtn[(2, 2)].sqrt()
    );

    // Physical interpretation
    println!("\n--- Physical Interpretation ---");
    println!("RTN frame aligns with the orbit:");
    println!("  R (Radial): Points from Earth center to satellite");
    println!("  T (Tangential): Points along velocity direction");
    println!("  N (Normal): Completes right-hand system (cross-track)");
    println!();
    println!("Key insight: Along-track (T) uncertainty grows fastest because");
    println!("velocity uncertainty causes timing errors that accumulate.");
    println!(
        "After one orbit: T/R ratio = {:.1}x",
        p_rtn[(1, 1)].sqrt() / p_rtn[(0, 0)].sqrt()
    );

    // Show correlation structure
    println!("\n--- Position Correlation Matrix (RTN) ---");
    let std_r = p_rtn[(0, 0)].sqrt();
    let std_t = p_rtn[(1, 1)].sqrt();
    let std_n = p_rtn[(2, 2)].sqrt();

    println!("       R      T      N");
    println!(
        "  R  {:6.3} {:6.3} {:6.3}",
        p_rtn[(0, 0)] / (std_r * std_r),
        p_rtn[(0, 1)] / (std_r * std_t),
        p_rtn[(0, 2)] / (std_r * std_n)
    );
    println!(
        "  T  {:6.3} {:6.3} {:6.3}",
        p_rtn[(1, 0)] / (std_t * std_r),
        p_rtn[(1, 1)] / (std_t * std_t),
        p_rtn[(1, 2)] / (std_t * std_n)
    );
    println!(
        "  N  {:6.3} {:6.3} {:6.3}",
        p_rtn[(2, 0)] / (std_n * std_r),
        p_rtn[(2, 1)] / (std_n * std_t),
        p_rtn[(2, 2)] / (std_n * std_n)
    );

    // Validate
    assert_eq!(p_gcrf.nrows(), 6);
    assert_eq!(p_rtn.nrows(), 6);
    assert!(p_rtn[(1, 1)].sqrt() > p_rtn[(0, 0)].sqrt()); // T > R

    println!("\nExample validated successfully!");
}
